// mcp_cp/CheckFactory.js
var feo = require("mcp_feo").feo;
var dlt = require("mcp_dlt").dlt;
var esf = require("mcp_esf").esf;
var kt = require("mcp_kt").kt;
var jc = require("mcp_jc").jc;
var jclq = require("mcp_jclq").jclq;
var pl3 = require("mcp_pl3").pl3;
var pl5 = require("mcp_pl5").pl5;
var sxc = require("mcp_sxc").sxc;

var CheckFactory = function(){};

CheckFactory.prototype.getCheck = function(gameCode, gameGrade, wNum) {
    var self = this;
    var gls = [];
    for(var key in gameGrade)
    {
        gls[gls.length] = gameGrade[key].bonus;
    }
    if(gameCode == 'T06')
    {
        return self.getCheckT06(gls, wNum)
    }
    else if(gameCode == 'T01')
    {
        return self.getCheckT01(gls, wNum)
    }
    else if(gameCode == 'T05')
    {
        return self.getCheckT05(gls, wNum)
    }
    else if(gameCode == 'F04')
    {
        return self.getCheckF04(gls, wNum)
    }
    else if(gameCode == 'T02')
    {
        return self.getCheckT02(gls, wNum)
    }
    else if(gameCode == 'T03')
    {
        return self.getCheckT03(gls, wNum)
    }
    else if(gameCode == 'T04')
    {
        return self.getCheckT04(gls, wNum)
    }
    else if(gameCode == 'T51') {
        return self.getCheckT51(wNum);

    }else if(gameCode == 'T52')
    {
        return self.getCheckT52(wNum);
    }
    else
    {
        return null;
    }
}

CheckFactory.prototype.getCheckT06 = function(gls, wNum)
{
    var self = this;
    var gl = feo.gl(gls.length);
    gl.setBonus(gls);
    var check = feo.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT01 =  function(gls, wNum)
{
    var self = this;
    var gl = dlt.gl(gls.length);
    gl.setBonus(gls);
    var check = dlt.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT05 =  function(gls, wNum)
{
    var self = this;
    var gl = esf.gl(gls.length);
    gl.setBonus(gls);
    var check = esf.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckF04 =  function(gls, wNum)
{
    var self = this;
    var gl = kt.gl(gls.length);
    gl.setBonus(gls);
    var check = kt.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT02 =  function(gls, wNum)
{
    var self = this;
    var gl = sxc.gl(gls.length);
    gl.setBonus(gls);
    var check = sxc.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT03 =  function(gls, wNum)
{
    var self = this;
    var gl = pl3.gl(gls.length);
    gl.setBonus(gls);
    var check = pl3.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT04 =  function(gls, wNum)
{
    var self = this;
    var gl = pl5.gl(gls.length);
    gl.setBonus(gls);
    var check = pl5.check();
    check.setDrawNum(wNum);
    check.setGl(gl);
    return check;
}

CheckFactory.prototype.getCheckT51 =  function(gls, wNum)
{
    var self = this;
    var check = jc.check();
    return check;
}

CheckFactory.prototype.getCheckT52 =  function(gls, wNum)
{
    var self = this;
    var check = jclq.check();
    return check;
}


module.exports = new CheckFactory();
