// mcp_control/ChannelQueryControl.js
var async = require('async');
var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;
var dateUtil = esut.dateUtil;

var dc = require('mcp_db').dc;

var cons = require('mcp_constants');
var userType = cons.userType;
var msgStatus = cons.msgStatus;
var msgType = cons.msgType;
var ticketPrintQueenStatus = cons.ticketPrintQueenStatus;
var ticketPrintStatus = cons.ticketPrintStatus;
var termStatus = cons.termStatus;
var gameGrade = cons.gameGrade;
var gameType = cons.gameType;


var config = require('mcp_config');
var ec = config.ec;
var prop = config.prop;
var game = config.game;


var service = require("mcp_service");
var digestSer = service.digestSer;
var moneySer = service.moneySer;
var kvSer = service.kvSer;
var ticketSer = service.ticketSer;
var notifySer = service.notifySer;

var ChannelQueryControl = function(){
    var self = this;
    self.cmd = {'CQ01':{ code:'CQ01', fromType:prop.digestFromType.FIX, des:"期次查詢"},
        'CQ02':{code:'CQ02', fromType:prop.digestFromType.FIX, des:'訂單查詢'},
        'CQ03':{code:'CQ03', fromType:prop.digestFromType.FIX, des:"票據查詢"},
        'CQ04':{code:'CQ04', fromType:prop.digestFromType.FIX, des:"奖级查询"},
        'CQ05':{code:'CQ05', fromType:prop.digestFromType.FIX, des:"期次报表查询"},
        'CQ06':{code:'CQ06', fromType:prop.digestFromType.FIX, des:"账户余额"},
        'CQ10':{code:'CQ10', fromType:prop.digestFromType.FIX, des:"查询系统时间"},
        'CQ22':{code:'CQ22', fromType:prop.digestFromType.FIX, des:"查询竞猜对阵信息"},
    };
    self.pTypeToKey = {"T5101":"hhad", "T5102":"had", "T5103":"crs", "T5104":"ttg", "T5105":"hafu", "T5201":"hdc", "T5202":"mnl", "T5203":"wnm", "T5204":"hilo" };
};

ChannelQueryControl.prototype.handle = function(headNode, bodyStr, userCb)
{
    var self = this;
    async.waterfall([
        //是否是支持的cmd
        function(cb)
        {
            var cmd = self.cmd[headNode.cmd];
            if(cmd == undefined)
            {
                cb(ec.E9000);
            }
            else
            {
                cb(null);
            }
        },
        //校验头的用户类型是否合法
        function(cb)
        {
            var userTypeId = userType[headNode.userType];
            if(userTypeId == undefined)
            {
                cb(ec.E9005);
            }
            else
            {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function(userTypeId, cb)
        {
            var cmd = self.cmd[headNode.cmd];
            digestSer.getKey({fromType:cmd.fromType, userId:headNode.userId, userType:userTypeId},
            function(err, key){
                cb(err, key);
            });
        },
        //先解密
        function(key, cb)
        {
            log.info(key);
            var decodedBodyStr = digestUtil.check(headNode, key, bodyStr);
            if(decodedBodyStr == null)
            {
                cb(ec.E9003);
                return;
            }
            var bodyNode;
            try {
                bodyNode = JSON.parse(decodedBodyStr);
                headNode.key = key;
            }
            catch (err)
            {
                cb(ec.E9003);
                return;
            }
            cb(null, bodyNode);
        },
        //check the param
        function(bodyNode, cb){
            var method = 'check' + headNode.cmd;
            self[method](null, headNode, bodyNode, function(err){
                cb(err, bodyNode);
            });
        },
        //业务处理
        function(bodyNode, cb){
            var cmd = 'handle' + headNode.cmd;
            self[cmd](null, headNode, bodyNode, cb);
        }
    ], function (err, bodyNode) {
        userCb(err, bodyNode);
    });
};


ChannelQueryControl.prototype.checkCQ01 = function(user, headNode, bodyNode, cb)
{
    if(bodyNode.gameCode == undefined || bodyNode.gameCode == null){
        cb(ec.E2070);
    }else{
        cb(null);
    }
};

ChannelQueryControl.prototype.checkCQ02 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

ChannelQueryControl.prototype.checkCQ03 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    cb(null);
};

ChannelQueryControl.prototype.checkCQ04 = function(user, headNode, bodyNode, cb)
{
    if(bodyNode.gameCode == undefined || bodyNode.gameCode == null){
        cb(ec.E2070);
    }else{
        cb(null);
    }
};

ChannelQueryControl.prototype.checkCQ05 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

ChannelQueryControl.prototype.checkCQ06 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

ChannelQueryControl.prototype.checkCQ07 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

ChannelQueryControl.prototype.checkCQ08 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


ChannelQueryControl.prototype.checkCQ09 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


ChannelQueryControl.prototype.checkCQ10 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

ChannelQueryControl.prototype.checkCQ22 = function(user, headNode, bodyNode, cb)
{
    if(bodyNode.passType != undefined && bodyNode.passType != null && bodyNode.passType != ""){
        bodyNode.single = bodyNode.passType;
    }else
    if(bodyNode.gameCode == undefined || bodyNode.gameCode == null || bodyNode.gameCode == ""){
        cb(ec.E2070);
        return;
    }
    if(bodyNode.matchCode == undefined || bodyNode.matchCode == null || bodyNode.matchCode == ""){
        delete  bodyNode.matchCode;
    }
    if(bodyNode.pType == undefined || bodyNode.pType == null || bodyNode.pType == ""){
        delete  bodyNode.pType;
    }else{
        if(bodyNode.pType instanceof Array){
            if(bodyNode.gameCode == "T51" && bodyNode.pType.length == 1 && (bodyNode.pType[0] == 6 || bodyNode.pType[0] == "06")){
                bodyNode.pType = ["01","02","03","04","05"];
                log.info("06玩法");
                cb(null);
                return;
            }else if(bodyNode.gameCode == "T52" && bodyNode.pType.length == 1 && (bodyNode.pType[0] == 5 || bodyNode.pType[0] == "05")){
                cb(null);
                return;
            }else{
                cb(null);
                return;
            }
        }else{
            cb(ec.E2070);
            return;
        }
    }
    cb(null);
};


/**
 * 期次查询
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ22 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {rst:[]};
    var table = dc.main.get("term");
    var cond = {};
    cond.gameCode = bodyNode.gameCode;
    if(bodyNode.matchCode){
        cond.code = bodyNode.matchCode;
    }else{
        cond.status = termStatus.ON_SALE;
    }
    var cols = { gameCode:1, code:1,  openTime:1, closeTime:1,};

    async.waterfall([
        function(cb)
        {
            var cursor = table.find(cond, cols, []).sort({openTime:1});
            cursor.dateToString();
            cursor.toArray(function(err, data){
               //查询的期次为竞猜期次。去查询赔率和明细
                async.eachSeries(data, function (row, callback) {
                    var jcoddsTable = dc.mg.get("jcodds");
                    var ticketGame = game.getInfo(row.gameCode);
                    row.matchCode = row.code;
                    if(ticketGame.type == gameType.Jingcai ){
                        var cond = {};
                        cond._id = row.gameCode +"_"+ row.code;
                        var fields = {l_cn:1,home_cn:1,guest_cn:1};
                        var jiaoyanfields = [];
                        if(bodyNode.pType){
                            for(var iterm in bodyNode.pType){
                                var tempType = bodyNode.pType[iterm].toString();
                                if(tempType.length == 1 && tempType < 10){
                                   tempType = "0"+tempType;
                                }
                                var key = self.pTypeToKey[bodyNode.gameCode+tempType];
                                fields[key] = 1;
                                jiaoyanfields.push(key);
                            }
                        }
                        jcoddsTable.findOne(cond, fields, {}, function (err, odds) {
                            if(odds != null && odds){
                                var undefindLength = 0;
                                for(var i = 0; i < jiaoyanfields.length; i++){
                                    if(odds[jiaoyanfields]){
                                        break;
                                    }else{
                                        undefindLength ++;
                                    }
                                }
                                if(undefindLength == jiaoyanfields.length){
                                    callback(null);
                                }else{
                                    for(var key in odds){
                                        row[key] = odds[key];
                                    }
                                    delete row._id;
                                    delete row.code;
                                    backBodyNode.rst.push(row);
                                    callback(null);
                                }
                            }else{
                                backBodyNode.rst.push(row);
                                callback(null);
                            }
                        });
                    }else{
                        callback(null);
                    }
                }, function (err) {
                    backBodyNode.count = cursor.count(function(err, count){
                        backBodyNode.count = count;
                        cb(err, null);
                    });
                });
            });
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}

/**
 * 期次查询
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ01 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.main.get("term");
    var cond = {};
    cond.gameCode = bodyNode.gameCode;
    if(bodyNode.termCode == undefined || bodyNode.termCode == null || bodyNode.termCode == ""){
        cond.status = termStatus.ON_SALE;
    }else{
        cond.code = bodyNode.termCode;
    }
    var cols = { gameCode:1, code:1,  openTime:1, closeTime:1, bonusTime:1, status:1 };
    async.waterfall([
        function(cb)
        {
            table.findOne(cond,cols, [], function(err, data){
                if(err){
                    cb(err);
                }else{
                    if(data){
                        data.termCode = data.code;
                        delete data.code;
                        backBodyNode =  data;
                        cb(null, backBodyNode);
                    }else{
                        cb(null, backBodyNode);
                    }
                }
            },{dateToString:true});
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}

/**
 * 订单查询
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ02 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.main.get("torder");
    var skip = bodyNode.skip;
    var limit = bodyNode.limit;
    pageUtil.parsePost(bodyNode);
    bodyNode.cond.customerId = headNode.userId;
    log.info(bodyNode);
    var cols = {id:1, amount:1, trueAmount:1, outerId:1, createTime:1};
    async.waterfall([
        function(cb)
        {
            var cursor = table.find(bodyNode.cond, cols, []).sort(bodyNode.sort).limit(bodyNode.skip, bodyNode.limit);
            cursor.dateToString();
            cursor.toArray(function(err, data){
                backBodyNode.rst = data;
                if(skip || limit){
                    backBodyNode.count = cursor.count(function(err, count){
                        backBodyNode.count = count;
                        cb(err, null);
                    });
                }else{
                    cb(err, null);
                }
            });
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}

/**
 * 票据查询
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ03 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.main.get("tticket");
    var skip = bodyNode.skip;
    var limit = bodyNode.limit;
    pageUtil.parsePost(bodyNode);
    bodyNode.cond.customerId = headNode.userId;

    var cols = {id:1, outerId:1, gameCode:1, termCode:1, pType:1, bType:1,
    createTime:1, status:1, printStatus:1, multiple:1, amount:1, bonus:1,
    bonusBeforeTax:1, bonusDetail:1,bonusStatus:1, number:1, dNumber:1, rNumber:1, orderId:1,
    province:1, seq:1, terminal:1, printId:1,printTime:1,cashTime:1,prizeTime:1,auditTime:1};
    async.waterfall([
        function(cb)
        {
            var cursor = table.find(bodyNode.cond, cols, []).sort(bodyNode.sort).limit(bodyNode.skip, bodyNode.limit);
            cursor.dateToString();
            cursor.toArray(function(err, data){
                backBodyNode.rst = data;
                if(skip || limit){
                    backBodyNode.count = cursor.count(function(err, count){
                        backBodyNode.count = count;
                        cb(err, null);
                    });
                }else{
                    cb(err, null);
                }
            });
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}

/**
 * 查询期次的奖级
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ04 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    pageUtil.parsePost(bodyNode);
    var skip = bodyNode.skip;
    var limit = bodyNode.limit;
    //指定了期次号，则从数据库中查询
    if(bodyNode.termCode)
    {
        var cols = {name:1, bonus:1, count:1, level:1};
        var table = dc.main.get("gamegrade");
        var termTable = dc.main.get("term");
        async.waterfall([
            function(cb){
                var cond =  {code: bodyNode.termCode, gameCode:bodyNode.gameCode};
                var cols = {code:1,gameCode:1,pool:1,totalsalemoney:1,wNum:1};
                termTable.findOne(cond, cols, [], function(err, data){
                     if(err){
                         cb(err);
                     }else{
                         if(data){
                             data.termCode = data.code;
                             delete  data.code;
                             backBodyNode = data;
                             cb(null);
                         }else{
                             cb(null);
                         }

                     }
                });
            },
            function(cb)
            {

                var cursor = table.find({gameCode:bodyNode.gameCode, termCode: bodyNode.termCode}, cols, []).sort(bodyNode.sort).limit(bodyNode.skip, bodyNode.limit);
                cursor.dateToString();
                cursor.toArray(function(err, data){
                    backBodyNode.bonusInfo = data;
                    if(skip || limit){
                        backBodyNode.count = cursor.count(function(err, count){
                            backBodyNode.count = count;
                            cb(err, null);
                        });
                    }else{
                        cb(err, null);
                    }
                });
            }
        ], function (err, rst) {
            cb(err, backBodyNode);
        });
    }
    //未指定期次号，从配置中读取
    else
    {
        var set = gameGrade.getInfoById(bodyNode.gameCode);
        if(set)
        {
            var rst = [];
            for(var key in set.grades)
            {
                var grade = set.grades[key];
                rst[rst.length] = {level:grade.id, name:grade.des, count:0,
                    gameCode:bodyNode.gameCode, bonus:grade.bonus, termCode:''};
            }
            backBodyNode.rst = rst;
        }
        cb(null, backBodyNode);
    }
}

/**
 * 期次报表查询
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ05 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var skip = bodyNode.skip;
    var limit = bodyNode.limit;
    pageUtil.parsePost(bodyNode);

    bodyNode.cond.customerId = headNode.userId;

    var cols = {
        gameCode:1, termCode:1, type:1,saleCount:1, saleAmount:1, hitCount:1, hitAmount:1, bonus:1,
        bonusBeforeTax:1, failCount:1, failAmount:1, createTime:1
    };
    var table = dc.main.get("termreport");
    async.waterfall([
        function(cb)
        {
            var cursor = table.find(bodyNode.cond, cols, []).sort(bodyNode.sort).limit(bodyNode.skip, bodyNode.limit);
            cursor.dateToString();
            cursor.toArray(function(err, data){
                backBodyNode.rst = data;
                if(skip || limit){
                    backBodyNode.count = cursor.count(function(err, count){
                        backBodyNode.count = count;
                        cb(err, null);
                    });
                }else{
                    cb(err, null);
                }
            });
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });

};

/**
 * 查询期次的奖级
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ06 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.mg.get("account");
    pageUtil.parsePost(bodyNode);
    bodyNode.cond.customerId = headNode.userId;

    var cols = {balance:1};
    async.waterfall([
        function(cb)
        {
            var cursor = table.findOne({"_id": headNode.userId}, cols, [], function(err ,data){
                if (err){
                    cb(err);
                }else{
                    delete data._id;
                    backBodyNode = data;
                    cb(err, null);
                }
            });
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}


/**
 * 查询时间 方便调整
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
ChannelQueryControl.prototype.handleCQ10 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.mg.get("account");
    pageUtil.parsePost(bodyNode);
    var requestTime = bodyNode.requestTime;
    async.waterfall([
        function(cb)
        {
             var dateTime = dateUtil.getCurTime();
            backBodyNode.requestTime = requestTime;
            backBodyNode.dateTime = dateTime;
            cb(null, backBodyNode);
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
}

module.exports = new ChannelQueryControl();
