// mcp_control/SalePageControl.js
var async = require('async');
var moment = require('moment');
var dc = require('mcp_db').dc;
var esdb = require('easy_db');
var config = require('mcp_config');
var prop = config.prop;
var game = config.game;
var ec = config.ec;
var ac = config.ac;

var esut = require('easy_util');
var dateUtil = esut.dateUtil;
var digestUtil = esut.digestUtil;
var log = esut.log;
var pageUtil = esut.pageUtil;

var util = require('mcp_util');
var dateMathUtil=util.dateMathUtil;
var excelUtil=util.excelUtil;

var service = require("mcp_service");
var digestSer = service.digestSer;

var cons = require('mcp_constants');
var ticketPrintStatus = cons.ticketPrintStatus;
var ticketStatus = cons.ticketStatus;
var userType = cons.userType;
var errUrl = cons.errUrl;
var accountSelect = cons.accountSelect;
var termReportType = cons.termReportType;

var SalePageControl = function () {
};

SalePageControl.prototype.handle = function (headNode, bodyNode, cb) {
    var self = this;
    var cmd = headNode.cmd;
    async.waterfall([
        //校验是否结束会话
        function (cb) {
            var userId = headNode.userId;
            if(userId == undefined)
            {
                if(headNode.userType!=undefined){
                    var userTypeId = userType[headNode.userType];
                    ec.E9006.url=errUrl.getInfoById(userTypeId).url;
                }
                cb(ec.E9006);
            }else{
                cb(null);
            }
        },
        //校验头的用户类型是否合法
        function(cb)
        {
            var userTypeId = userType[headNode.userType];
            if(userTypeId == undefined)
            {
                cb(ec.E9005);
            }
            else
            {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function(userTypeId, cb)
        {
            var fromType = prop.digestFromType.CACHE;
            digestSer.getKey({fromType:fromType, userId:headNode.userId, userType:userTypeId},
                function(err, key){
                    cb(err, key);
                });
        },
        //校验密钥并处理逻辑
        function(key, cb)
        {
            if(key != headNode.key)
            {
                cb(ec.E9003);
            }
            else
            {
                self[cmd[1]](headNode, bodyNode, cb);
            }
        }
    ], function (err, bodyNode) {
        cb(err, bodyNode);
    });
};


SalePageControl.prototype.term = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {title: "view terms"};
    backBodyNode.sort = {id: -1};
    pageUtil.parse(bodyNode, backBodyNode);
    var table = dc.main.get("termreport");
    backBodyNode.cond.customerId = headNode.userId;
    backBodyNode.game = game.getInfo();
    var add = backBodyNode.add;
    if (add.startDate != null && add.endDate != null && add.startDate != "" && add.endDate != "") {
        add.startDate = dateMathUtil.getEarlyTime(add.startDate);
        add.endDate =  dateMathUtil.getLaterTime(add.endDate);
        backBodyNode.cond.createTime = {$gte: add.startDate, $lte: add.endDate};
        //页面显示时间格式
        add.startDate = moment(add.startDate).format("YYYY-MM-DD");
        add.endDate =  moment(add.endDate).format("YYYY-MM-DD");
    }
    backBodyNode.cond.type = termReportType['SALE'];
    var cursor = table.find(backBodyNode.cond, {}, []).sort(backBodyNode.sort).limit(backBodyNode.skip, backBodyNode.limit);
    cursor.toArray(function (err, data) {
        for (var key in data) {
            var set = data[key];
            set.createTime = dateUtil.toString(set.createTime);
        }
        backBodyNode.rst = data;
        backBodyNode.count = cursor.count(function (err, count) {
            backBodyNode.count = count;
            cb(null, backBodyNode);
        });
    });
}


SalePageControl.prototype.day = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {title: "view day"};
    //获取游戏种类
    backBodyNode.game = game.getInfo();
    pageUtil.parse(bodyNode, backBodyNode);
    //时间条件
    var add = backBodyNode.add;
    backBodyNode.rst = {};
    backBodyNode.count = 0;
    if (add.startDate != null && add.endDate != null && add.startDate != "" && add.endDate != "") {
        if(dateMathUtil.isValidDate(add.startDate) && dateMathUtil.isValidDate(add.endDate)){
            if(dateMathUtil.daysBetween(add.startDate,add.endDate)>31){
                backBodyNode.err="查询时间不能超过一个月！";
                cb(null,backBodyNode);
                return;
            }else{
                add.startDate = dateMathUtil.getEarlyTime(add.startDate);
                add.endDate =  dateMathUtil.getLaterTime(add.endDate);
            }
        }else{
            backBodyNode.err="请注意时间格式！";
            cb(null,backBodyNode);
            return;
        }
    }else{
        add.startDate = dateMathUtil.getEarlyTime();
        add.endDate =  dateMathUtil.getLaterTime();
    }
    var addCond = "";
    addCond = "and createTime>=" + dateMathUtil.getTimestamp(add.startDate) + " and createTime<=" + dateMathUtil.getTimestamp(add.endDate);
    //页面显示时间格式
    add.startDate = moment(add.startDate).format("YYYY-MM-DD");
    add.endDate =  moment(add.endDate).format("YYYY-MM-DD");

    //请求条件
    backBodyNode.cond.customerId = headNode.userId;
    var reqDate = backBodyNode.cond;
    var cond = "";
    for (var key in reqDate) {
        cond += "and " + key + "='" + reqDate[key]+"'";
    }
    //分页
    var pageCond="limit "+backBodyNode.skip+","+backBodyNode.limit;

    //链接查询
    var conn = dc.main.getConn();
    if (dc.main.type == esdb.prop.dbType.mysql) {
        var sql = "select gameCode,count(case when printStatus = 1300 then 1 else null end ) as salenum,sum(case when printStatus = 1300 then amount else 0 end) as saleamount,count(case when status = 1200 then 1 else null end ) as prizenum,sum(case when status = 1200 then amount else 0 end ) as prizeamount,count(case when status = 1400 then 1 else null end ) as refundnum,sum(case when status = 1400 then amount else 0 end ) as refundamount from tticket A where 1=1 " + cond +"  "+ addCond+" group by gamecode  "+pageCond;
        console.log("sql: "+sql);
        conn.execute(sql, {}, function (err, data) {
            if(err){
                console.log(err);
            }else{
            backBodyNode.rst = data;
            backBodyNode.count = data.length;
            cb(null, backBodyNode);
            }
        });
    } else {

        //其他数据类型

    }
}


SalePageControl.prototype.print = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {title: "view print"};
    pageUtil.parse(bodyNode, backBodyNode);
    backBodyNode.cond.customerId = headNode.userId;
    backBodyNode.game = game.getInfo();
    backBodyNode.ticketPrintStatus = ticketPrintStatus.getInfoById();
    backBodyNode.prizeStatus = [{"id":ticketStatus.HIT, des: "是"},{"id":ticketStatus.NOT_HIT, des:"否"}];

    backBodyNode.sort={id: -1};
    var add = backBodyNode.add;
    if (add.startDate != null && add.endDate != null && add.startDate != "" && add.endDate != "") {
        add.startDate = dateMathUtil.getEarlyTime(add.startDate);
        add.endDate =  dateMathUtil.getLaterTime(add.endDate);
        backBodyNode.cond.createTime = {$gte: add.startDate, $lte: add.endDate};
        //页面显示时间格式
        add.startDate = moment(add.startDate).format("YYYY-MM-DD");
        add.endDate =  moment(add.endDate).format("YYYY-MM-DD");
    }
    if (add.printStartDate != null && add.printEndDate != null && add.printStartDate != "" && add.printEndDate != "") {
        add.printStartDate = dateMathUtil.getEarlyTime(add.printStartDate);
        add.printEndDate =  dateMathUtil.getLaterTime(add.printEndDate);
        backBodyNode.cond.printTime = {$gte: add.printStartDate, $lte: add.printEndDate};
        //页面显示时间格式
        add.printStartDate = moment(add.printStartDate).format("YYYY-MM-DD");
        add.printEndDate =  moment(add.printEndDate).format("YYYY-MM-DD");
    }
    async.waterfall([
        function (cb) {
            var table = dc.main.get("tticket");
            var cursor = table.find(backBodyNode.cond, {}, []).sort(backBodyNode.sort).limit(backBodyNode.skip, backBodyNode.limit);
            cursor.dateToString();
            cursor.toArray(function (err, data) {
                for (var key in data) {
                    var set = data[key];
                    set.game = game.getInfo(set.gameCode);
                    set.printStatus = ticketPrintStatus.getInfoById(set.printStatus);
                }
                backBodyNode.rst = data;
                backBodyNode.count = cursor.count(function (err, count) {
                    backBodyNode.count = count;
                    cb(null, data);
                });
            });
        }, function (ttickets, cb) {
            async.each(ttickets, function (row, callback) {
                var table = dc.main.get("torder");
                table.findOne({id: row.orderId}, {}, [], function (err, data) {
                    if (err) {
                        callback(ec.E0999);
                    }
                    else {
                        if (data) {
                            row.orderId = data.outerId;
                        }
                        callback();
                    }
                });
            }, function (err) {
                cb(err);
            })
        },
        function(cb) {
            //小计
            var table = dc.main.get("tticket");
            var sqlCondition = table.condition(backBodyNode.cond);
            var sql = "select count(*) as num, sum(amount) as amount, sum(bonus) as bonus from tticket where " + sqlCondition;
            log.info(sql);
            var conn = dc.main.getConn();
            backBodyNode.num = 0;
            backBodyNode.amount = 0;
            backBodyNode.bonus = 0;
            conn.execute(sql, [], function (err, data) {
                if (err) {
                    log.error(err);
                }else{
                    if(data){
                        backBodyNode.num = data[0].num;
                        backBodyNode.amount = data[0].amount;
                        backBodyNode.bonus = data[0].bonus;
                        cb(null);
                    }
                }
            });
        }
    ], function (err) {
        cb(err, backBodyNode);
    })

}

SalePageControl.prototype.download = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    pageUtil.parse(bodyNode, backBodyNode);
    backBodyNode.cond.customerId = headNode.userId;
    backBodyNode.game = game.getInfo();
    backBodyNode.ticketPrintStatus = ticketPrintStatus.getInfoById();
    backBodyNode.prizeStatus = [{"id":ticketStatus.HIT, des: "是"},{"id":ticketStatus.NOT_HIT, des:"否"}];

    backBodyNode.sort={id: -1};
    var add = backBodyNode.add;
    if (add.startDate != null && add.endDate != null && add.startDate != "" && add.endDate != "") {
        add.startDate = dateMathUtil.getEarlyTime(add.startDate);
        add.endDate =  dateMathUtil.getLaterTime(add.endDate);
        backBodyNode.cond.createTime = {$gte: add.startDate, $lte: add.endDate};
        //页面显示时间格式
        add.startDate = moment(add.startDate).format("YYYY-MM-DD");
        add.endDate =  moment(add.endDate).format("YYYY-MM-DD");
    }
    if (add.printStartDate != null && add.printEndDate != null && add.printStartDate != "" && add.printEndDate != "") {
        add.printStartDate = dateMathUtil.getEarlyTime(add.printStartDate);
        add.printEndDate =  dateMathUtil.getLaterTime(add.printEndDate);
        backBodyNode.cond.printTime = {$gte: add.printStartDate, $lte: add.printEndDate};
        //页面显示时间格式
        add.printStartDate = moment(add.printStartDate).format("YYYY-MM-DD");
        add.printEndDate =  moment(add.printEndDate).format("YYYY-MM-DD");
    }
    var conf = {};
    conf.cols = [{caption:"出票时间",type:'string'},{caption:"彩票Id",type:'string'},{caption:"出票状态",type:'string'},
        {caption:"票状态",type:'string'},{caption:"彩种",type:'string'},{caption:"金额",type:'string'},{caption:"中奖金额",type:'string'}];
    conf.rows = new Array();
    var fields = {printTime:1,outerId:1,printStatus:1,status:1,gameCode:1,amount:1,bonus:1};
    async.waterfall([
        function (cb) {
            var table = dc.main.get("tticket");
            var cursor = table.find(backBodyNode.cond, fields, []).sort(backBodyNode.sort);
            cursor.dateToString();
            cursor.toArray(function (err, data) {
                for (var key in data) {
                    var set = data[key];
                    set.gameCode = game.getInfo(set.gameCode);
                    set.printStatus = ticketPrintStatus.getInfoById(set.printStatus);
                    set.status = ticketStatus.getInfoByCode(set.status);
                    rows.push(set.printTime);
                    rows.push(set.outerId);
                    rows.push(set.printStatus);
                    rows.push(set.status);
                    rows.push(set.gameCode);
                    rows.push(set.amount);
                    rows.push(set.bonus);
                }
                cb(null);
            });
        }, function (cb) {
            excelUtil.createExcel(conf, function(err, filePath){
                log.info("创建完成"+filePath);
                cb(null, filePath);
            })
        }
    ], function (err, filePath) {
        log.info("路径"+filePath);
        cb(err, filePath);
    })

}

SalePageControl.prototype.money = function (headNode, bodyNode, cb) {

    var self = this;
    var backBodyNode = {title:"账户流水查询"};
    if(bodyNode.sort == undefined)
    {
        backBodyNode.sort = {createTime:-1};
    }
    pageUtil.parse(bodyNode, backBodyNode);
    var table = dc.main.get("moneylog");
    backBodyNode.cond.customerId = headNode.userId;

    //页面显示科目查询项
    var selectValues=accountSelect.getInfoByCode(headNode.userType);
    backBodyNode.selectValues=JSON.stringify(selectValues.accounts);

    var add = backBodyNode.add;
    if (add.startDate != null && add.endDate != null && add.startDate != "" && add.endDate != "") {
        add.startDate = dateMathUtil.getEarlyTime(add.startDate);
        add.endDate =  dateMathUtil.getLaterTime(add.endDate);
        backBodyNode.cond.createTime = {$gte: add.startDate, $lte: add.endDate};
        //页面显示时间格式
        add.startDate = moment(add.startDate).format("YYYY-MM-DD");
        add.endDate =  moment(add.endDate).format("YYYY-MM-DD");
    }

    var sqlCondition = table.condition(backBodyNode.cond);
    var sql = "select sum(amount) as amount,typeId from moneylog where " + sqlCondition + " group by typeId";
    log.info(sql);
    var conn = dc.main.getConn();
    conn.execute(sql, [], function(err ,data){
        if(err)
        {
            log.error(err);
        }

        backBodyNode.inAmount = 0;
        backBodyNode.outAmount = 0;
        if(data)    //设置统计信息到界面
        {
            for(var i = 0; i < data.length; i++)
            {
                var set = data[i];
                if(set.typeId == '0')
                {
                    backBodyNode.inAmount = set.amount;
                }
                else if(set.typeId == '1')
                {
                    backBodyNode.outAmount = (-1)*set.amount;
                }
            }
        }

        console.log("moneylog:" +JSON.stringify(backBodyNode.cond.createTime));
        var cursor = table.find(backBodyNode.cond, {}, []).sort(backBodyNode.sort).limit(backBodyNode.skip, backBodyNode.limit);
        cursor.toArray(function(err, data){
            for(var key in data)
            {
                var set = data[key];
                set.subject = ac.getInfo(set.roleId, set.accountId, set.typeId, set.subjectId);
                set.createTime = moment(set.createTime).format("YYYY-MM-DD HH:mm:ss")
            }
            backBodyNode.rst = data;
            backBodyNode.count = cursor.count(function(err, count){
                backBodyNode.count = count;
                cb(null, backBodyNode);
            });
        });

    });
}

module.exports = new SalePageControl();
